lib/azuremetadata/azuremetadata.py

```diff
@@ -15,25 +22,58 @@ def get_all(self):
         return result
 
     def get_instance_data(self):
-        # FIXME this is a stub while the CLI UX is being figured out
-
-        if self._api_version == '2019-08-15':
-            with open('fixtures/metadata-v2019-08-15-multiple-network-if.json', 'r') as json_file:
-                return json.load(json_file)
-        elif self._api_version == '2017-04-02':
-            with open('fixtures/metadata-v2017-04-02.json', 'r') as json_file:
-                return json.load(json_file)
-        else:
-            raise Exception("Unknown API version fixture exception")
+        return self._make_request(
+            f"http://169.254.169.254/metadata/instance?api-version={quote(self._api_version)}"
+        )
 
     def get_attested_data(self):
-        # FIXME this is a stub while the CLI UX is being figured out
-
-        if self._api_version == '2019-08-15':
-            with open('fixtures/attested-data-v2019-08-15.json', 'r') as json_file:
-                return json.load(json_file)
-        elif self._api_version == '2018-10-01':
-            with open('fixtures/attested-data-v2018-10-01.json', 'r') as json_file:
-                return json.load(json_file)
-        else:
-            raise Exception("Unknown API version fixture exception (attested data)")
+        return self._make_request(
+            f"http://169.254.169.254/metadata/attested/document?api-version={quote(self._api_version)}"
+        )
+
+    def get_disk_tag(self, device=None):
+        if not device:
+            device = self._find_root_device()
+
+        with open(device, 'rb') as fh:
+            fh.seek(65536)
+            return str(uuid.UUID(bytes_le=fh.read(16)))
+
+    @staticmethod
+    def _find_root_device():
+        hex_device_id = os.stat("/").st_dev
+        root_device_id = str(os.major(hex_device_id)) + ":" + str(os.minor(hex_device_id))
+
+        devices = glob.glob("/sys/block/*")
+        for device_path in devices:
+            device = os.path.basename(device_path)
+            partitions = glob.glob(f"/sys/block/{device}/{device}*")
+
+            for partition in partitions:
+                with open(f"{partition}/dev", "r") as fh:
+                    device_id = fh.read().strip()
+
+                if device_id == root_device_id:
+                    return f"/dev/{device}"
+
+        raise Exception("Can't find root device")
+
+    @staticmethod
+    def _make_request(url):
+        tries = 0
+        last_error = None
+        while tries < 5:
+            try:
+                req = urllib.request.Request(url, headers={'Metadata': 'true'})
+                response = urllib.request.urlopen(req)
+
+                data = response.read()
+                return json.loads(data)
+            except urllib.error.HTTPError as e:
+                raise e
```

Comment:
Similar to previous comment. Best effort implementation should be used. HTTPError for one end point does not mean we'll get the same for another.

Reply:
@rjschwei 
1. Didn't we agree that if `--tag` is specified, then it wouldn't be appended to the rest of the metadata since it's not the part of it? `azuremetadata --tag --external-IP`  isn't possible anymore, if `--tag` is specified prints the just the tag and exits.
2. 
 * Handling `HTTPError` is mostly inspired by how `gcemetadata` handles exceptions ( consistency across different frameworks?):
   * https://github.com/SUSE-Enceladus/gcemetadata/blob/master/lib/gcemetadata/gcemetadata.py#L57
   * https://github.com/SUSE-Enceladus/gcemetadata/blob/master/lib/gcemetadata/gcemetadata.py#L61
 * Also I don't see how ignoring the errors could be useful:
   * It won't set the exit code, making the CLI tool unusable in scripts of any kind
   * If the data is unavailable -- it will break querying in all kinds of weird ways (i.e., it will complain that unknown parameters are specified, because if the data can't be fetched then those arguments couldn't be parsed properly)

Reply:
@ikapelyukhin Sorry for not being able to express myself better. `azuremetadata --tag --external-IP` needs to be possible, the user shouldn't have to know where the information is coming from. It is the tools responsibility to find the "right" source and then present the collected data to the user. Just like `azuremetadata --external-IP --attested-data` is/should be possible. The user doesn't care that the data comes from different endpoints.

We can build logic into the command line parser knows how to use the underlying utils for the different sources, or we can build logic that merges the information from underlying sources into 1 internal data structure. Either is fine as long as the user is not being forced to be aware of these implementation details.

Hope this clarifies the intended behavior and user experience.

Reply:
@ikapelyukhin In GCE there is no other data, i.e. the only source of information about the instance is the metadata server. Therefore if the metadata server cannot be reached https://github.com/SUSE-Enceladus/gcemetadata/blob/master/lib/gcemetadata/gcemetadata.py#L57 or if the API version https://github.com/SUSE-Enceladus/gcemetadata/blob/master/lib/gcemetadata/gcemetadata.py#L61 is set improperly we cannot query any data and thus need to exit. We cannot perform a best effort.

You can see here https://github.com/SUSE-Enceladus/gcemetadata/blob/master/lib/gcemetadata/gcemetadata.py#L166 that an error accessing the endpoint of the metadata server does not end the query operation, rather it implements a best effort approach.

What appears as an inconsistency https://github.com/SUSE-Enceladus/gcemetadata/blob/master/lib/gcemetadata/gcemetadata.py#L353 is not an inconsistency because set_disk_device() is only called from display_all() all where the IDs are generated and thus we'd never trigger that exception. That exception exists in case another developer wants to use the metadata class in different ways.

Hope this makes sense.

Reply:
@rjschwei OK, then what should happen if no parameters are specified? Should the tag be present in the output with the rest of the metadata or not?

Reply:
Given that the disk tag is somewhat of a "hidden" feature I am OK if we leave it out of the output of `azuremetadata` . But we have to remember to add it to the supportutils plugin as an extra command to execute.

Reply:
@rjschwei no, it actually makes more sense to always append it, then. `--tag` is going away, as the tag will be always present, `--billingTag` will be used to query just the billing tag. I'll just change the implementation to return an empty string if reading it has failed for some reason.

Reply:
@ikapelyukhin Also fine, but be aware for images where we do not specifically write the data the place where the tag is located probably has random data @schaefi does kiwi zero out the bits where the billing GUID is written or set it tom something obviously bogus like all 1s?

Meaning we will most likely always be able to read a disk tag, it's just going to meaningless, which is also OK.

Reply:
@rjschwei OK, now all failures are soft failures.